Zend/zend_alloc.c

```diff
@@ -3009,6 +3085,116 @@ ZEND_API void zend_mm_get_custom_handlers(zend_mm_heap *heap,
 #endif
 }
 
+/*
+ * Use this function to register observers to the ZendMM.
+ * This function operates on the thread local heap and is meant to be called
+ * during the request lifetime. Calling it in MINIT works only in NTS builds,
+ * but not in ZTS.
+ */
+ZEND_API zend_mm_observer* zend_mm_observer_register(
+	zend_mm_heap* heap,
+	void (*malloc)(size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
+	void (*free)(void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
+	void (*realloc)(void *, size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC)
+) {
+#if ZEND_MM_CUSTOM
+	if (!heap) {
+		heap = AG(mm_heap);
+	}
+	ZEND_ASSERT(heap);
+
+	zend_mm_observer *node = pemalloc(sizeof(zend_mm_observer), 1);
+	node->malloc = malloc;
+	node->free = free;
+	node->realloc = realloc;
+	node->next = NULL;
+
+	// set bitflag for observers being around
+	heap->use_custom_heap |= ZEND_MM_CUSTOM_HEAP_OBSERVED;
+
+	if (heap->observers == NULL) {
+		heap->observers = node;
+		return node;
+	}
+	zend_mm_observer *current = heap->observers;
+	while (current->next != NULL) {
+		current = current->next;
+	}
+	current->next = node;
+
+	return node;
+#else
+	return NULL;
+#endif
+}
+
+/*
+ * Use this function to unregister your observer at any given time. In case your
+ * observer was the last one to be unregistered, this will also reset the
+ * `heap->use_custom_heap` bitflag.
+ */
+ZEND_API bool zend_mm_observer_unregister(zend_mm_heap *heap, zend_mm_observer *observer)
+{
+#if ZEND_MM_CUSTOM
+	if (!heap) {
+		heap = AG(mm_heap);
+	}
+	ZEND_ASSERT(heap);
+
+	zend_mm_observer *current = heap->observers, *prev = NULL;
+
+	if (current == observer) {
```

Comment:
Nit: This is somewhat redundant with the below. We could put the `prev->next = current->next;` inside a `if (prev != NULL)` check with an `else { heap->observers = current->next; if (!current->next) { heap->use_custom_heap &= ~ZEND_MM_CUSTOM_HEAP_OBSERVED; } }`. I'm not sure if that's simpler without seeing it written out.

Reply:
I am not 100% sure I got your proposal right, but I reworked that unregister function and I think it is way clearer and simpler now.

Reply:
Looks perfect :slightly_smiling_face: